EmulatorDetectorProject/EmulatorDetector/src/main/java/net/skoumal/emulatordetector/EmulatorDetector.java

```diff
@@ -26,56 +29,126 @@
+                Build.MODEL.contains("TiantianVM") ||
+                Build.MODEL.contains("Andy") ||
+                Build.MODEL.equals("Android SDK built for x86_64") ||
+                Build.MODEL.equals("Android SDK built for x86")) {
+            rating++;
+        }
 
-            if(Build.DEVICE.equals("generic") ||
-                    Build.DEVICE.equals("generic_x86") ||
-                    Build.DEVICE.equals("vbox86p")) {
-                newRating ++;
-            }
+        if (Build.HARDWARE.equals("goldfish") ||
+                Build.HARDWARE.equals("vbox86") ||
+                Build.HARDWARE.contains("nox") ||
+                Build.HARDWARE.contains("ttVM_x86")) {
+            rating++;
+        }
 
-            if(Build.MODEL.equals("sdk") ||
-                    Build.MODEL.equals("google_sdk") ||
-                    Build.MODEL.equals("Android SDK built for x86")) {
-                newRating ++;
+        if (Build.FINGERPRINT.contains("generic/sdk/generic") ||
+                Build.FINGERPRINT.contains("generic_x86/sdk_x86/generic_x86") ||
+                Build.FINGERPRINT.contains("Andy") ||
+                Build.FINGERPRINT.contains("ttVM_Hdragon") ||
+                Build.FINGERPRINT.contains("generic_x86_64") ||
+                Build.FINGERPRINT.contains("generic/google_sdk/generic") ||
+                Build.FINGERPRINT.contains("vbox86p") ||
+                Build.FINGERPRINT.contains("generic/vbox86p/vbox86p")) {
+            rating++;
+        }
+
+        try {
+            String opengl = android.opengl.GLES20.glGetString(android.opengl.GLES20.GL_RENDERER);
+            if (opengl != null){
+                if( opengl.contains("Bluestacks") || 
+                    opengl.contains("Translator")
+                )
+                rating += 10;
             }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
 
-            if(Build.HARDWARE.equals("goldfish") ||
-                    Build.HARDWARE.equals("vbox86")) {
-                newRating ++;
+        try {
+            File sharedFolder = new File(Environment
+                    .getExternalStorageDirectory().toString()
+                    + File.separatorChar
+                    + "windows"
+                    + File.separatorChar
+                    + "BstSharedFolder");
+
+            if (sharedFolder.exists()) {
+                rating += 10;
             }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
 
-            if(Build.FINGERPRINT.contains("generic/sdk/generic") ||
-                    Build.FINGERPRINT.contains("generic_x86/sdk_x86/generic_x86") ||
-                    Build.FINGERPRINT.contains("generic/google_sdk/generic") ||
-                    Build.FINGERPRINT.contains("generic/vbox86p/vbox86p")) {
-                newRating ++;
+        try {
+            BluetoothAdapter m_BluetoothAdapter = BluetoothAdapter.getDefaultAdapter();

```

Comment:
You should handle BluetoothAdapter.getDefaultAdapter() vs. BluetoothAdapter.getAdapter() according to documentation: https://developer.android.com/reference/android/bluetooth/BluetoothAdapter.html

Actually I don't like this BT check at all, because it requires BT permissions. Should be applied only when particular permission is available.